flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/SubpartitionDiskCacheManager.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * The {@link SubpartitionDiskCacheManager} is responsible to manage the cached buffers in a single
+ * subpartition.
+ */
+class SubpartitionDiskCacheManager {
+
+    /**
+     * All the buffers. The first field of the tuple is the buffer, while the second field of the
+     * buffer is the buffer index.
+     *
+     * <p>Note that this field can be accessed by the task thread or the write IO thread, so the
+     * thread safety should be ensured.
+     */
+    private final Deque<Tuple2<Buffer, Integer>> allBuffers = new LinkedList<>();
+
+    /**
+     * Record the buffer index in the {@link SubpartitionDiskCacheManager}. Each time a new buffer
+     * is added to the {@code allBuffers}, this field is increased by one.
+     */
+    private int bufferIndex;
+
+    /**
+     * Record the segment id that is writing to. Each time when the segment is finished, this filed
+     * is increased by one.
+     */
+    private int segmentIndex;
+
+    // ------------------------------------------------------------------------
+    //  Called by DiskCacheManager
+    // ------------------------------------------------------------------------
+
+    void append(Buffer buffer) {
+        addBuffer(buffer);
+    }
+
+    void appendEndOfSegmentEvent(ByteBuffer record, DataType dataType) {
+        writeEvent(record, dataType);
+        segmentIndex++;
+    }
+
+    /** Note that allBuffers can be touched by multiple threads. */
+    List<Tuple2<Buffer, Integer>> removeAllBuffers() {
+        synchronized (allBuffers) {
+            List<Tuple2<Buffer, Integer>> targetBuffers = new ArrayList<>(allBuffers);
+            allBuffers.clear();
+            return targetBuffers;
+        }
+    }
+
+    int getBufferIndex() {
+        return bufferIndex;
+    }
+
+    int getSegmentIndex() {
+        return segmentIndex;
+    }
+
+    void release() {
+        recycleBuffers();
+        checkState(allBuffers.isEmpty(), "Leaking buffers.");
+    }
+
+    // ------------------------------------------------------------------------
+    //  Internal Methods
+    // ------------------------------------------------------------------------
+
+    private void writeEvent(ByteBuffer event, DataType dataType) {
+        checkArgument(dataType.isEvent());
+
+        MemorySegment data = MemorySegmentFactory.wrap(event.array());
+        addBuffer(new NetworkBuffer(data, FreeingBufferRecycler.INSTANCE, dataType, data.size()));
+    }
+
+    /** Note that allBuffers can be touched by multiple threads. */
```

Comment:
Is this method only called by task thread?

Reply:
Fixed.